### PR TITLE
Create separate compression-specific layer to enable writing gzipped files

### PR DESCRIPTION
**Pull request by mpenkov**
This PR addresses issue #91. There are two major parts to this PR.

First, I rewrote all the lower-level S3 stuff using boto3 in a separate module, smart_open/s3.py. he reader and writer inherit from the [io classes in the standard library](https://docs.python.org/3/library/io.html), meaning that you can pretty much treat them like local file streams and pass them to codecs like gzip, bz2, etc. 

Second, I integrated this module into smart_open_lib.py. This part required a rebase: I started working on this over a year ago, and much has happened since then. The part I couldn't handle well during the rebase was: there was a new `s3u` mode added, with code that depends on the old boto library. I couldn't port this part to boto3 - this functionality may be [discontinued](https://github.com/boto/boto3/issues/334). The code is isolated in the `setup_unsecured_mode` function and is currently commented out.

This takes care of #91 for S3 only. In the future, we should apply the same approach to other protocols like HDFS, more specifically: make sure readers and writers inherit from stdlib io classes.

### CODE REVIEW

**Comment by mpenkov** [comment]
I forgot to mention that this also resolves #42 (s3 seek), and partially resolves #43 (starts boto3 migration).

I'm not 100% sure, but it should also resolve #64 and #74.

**Comment by piskvorky** [comment]
Epic! @menshikh-iv let's review ASAP.

We should probably clean up and merge the smaller PRs first, this will be a major update.

@mpenkov what kind of backward incompatibilities do you expect here?

**Comment by mpenkov** [comment]
@piskvorky I don't really understand the s3u stuff that was added in the meanwhile. Since that's currently effectively disabled in this branch, there is a chance of an incompatibility there.

Other than that, I don't expect these changes to break anything - they still pass pretty much the same tests and work the same way.

File: smart_open/s3.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+import six
+
+
+_LOGGER = logging.getLogger(__name__)
+
+#
+# https://github.com/matthewwithanm/django-imagekit/commit/0d5bfe37517955cca2284735769fe1dffb38ed37
+#
+try:
```
Comment (menshikh-iv) [inline smart_open/s3.py]: Need to remove (we dropped 2.6 support)

File: smart_open/s3.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+"""Implements file-like objects for reading and writing from/to S3."""
+import boto3
+
+import io
+import logging
+
+import six
+
+
+_LOGGER = logging.getLogger(__name__)
```
Comment (menshikh-iv) [inline smart_open/s3.py]: please use `logger` (everywhere)

File: smart_open/s3.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+            raw = self._raw_reader.read(size=io.DEFAULT_BUFFER_SIZE)
+            if len(raw):
+                self._buffer += raw
+            else:
+                _LOGGER.debug('reached EOF while filling buffer')
+                self._eof = True
+
+        return self._read_from_buffer(size)
+
+    def read1(self, size=-1):
```
Comment (menshikh-iv) [inline smart_open/s3.py]: Why this needed here?

File: smart_open/s3.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        """This is the same as read()."""
+        return self.read(size=size)
+
+    def readinto(self, b):
```
Comment (menshikh-iv) [inline smart_open/s3.py]: Very strange, why we need this method with mutable `b`?

File: smart_open/s3.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        """Read up to len(b) bytes into b, and return the number of bytes
+        read."""
+        data = self.read(len(b))
+        if not data:
+            return 0
+        b[:len(data)] = data
+        return len(data)
+
+    def terminate(self):
+        """Do nothing."""
```
Comment (menshikh-iv) [inline smart_open/s3.py]: Add `pass`

File: smart_open/s3.py (hunk 3 above)
Comment (mpenkov) [reply smart_open/s3.py]: It's part of the interface for [BufferedIOBase](https://docs.python.org/2/library/io.html#io.BufferedIOBase.read1).

File: smart_open/s3.py (hunk 4 above)
Comment (mpenkov) [reply smart_open/s3.py]: Same as above. We try to behave like a [BufferedIOBase](https://docs.python.org/2/library/io.html#io.BufferedIOBase.readinto), which defines this method.

Some application downstream of smart_open may depend on this functionality.

**Comment by mpenkov** [comment]
@menshikh-iv I made changes in response to your review. I left two items as they are because they are required to satisfy interface contracts. Let me know if I need to do anything else. Thanks!

**Comment by menshikh-iv** [comment]
Thanks for clarification about `BufferIOBase`, great work, big thanks for your contribution @mpenkov :fire: :fire: :fire:

**Pull request merged**